messaging/src/main/java/com/flipkart/varadhi/produce/otel/ProducerMetricProvider.java

```diff
@@ -0,0 +1,56 @@
+package com.flipkart.varadhi.produce.otel;
+
+import com.flipkart.varadhi.entities.ProduceContext;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import static com.flipkart.varadhi.Constants.Meters.Produce.BYTES_METER;
+import static com.flipkart.varadhi.Constants.Meters.Produce.LATENCY_METER;
+import static com.flipkart.varadhi.Constants.Tags.*;
+
+public class ProducerMetricProvider {
+    private final MeterRegistry meterRegistry;
+    private boolean enabled;
+
+    public ProducerMetricProvider(boolean enabled, MeterRegistry meterRegistry) {
+        this.meterRegistry = meterRegistry;
+        this.enabled = enabled;
+    }
+
+    public void OnMessageProduced(boolean succeeded, long producerLatency, ProduceContext context) {
+        if (enabled) {
+            List<Tag> tags = getTags(context);
+            getProducedBytesCounter(tags).increment(context.getRequestContext().getBytesReceived());
+            tags.add(Tag.of(TAG_NAME_PRODUCE_RESULT, succeeded ? TAG_VALUE_RESULT_SUCCESS : TAG_VALUE_RESULT_FAILED));
+            getLatencyTimer(tags).record(producerLatency, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private List<Tag> getTags(ProduceContext context) {
+        List<Tag> tags = new ArrayList<>();
+        tags.add(Tag.of(TAG_NAME_REGION, context.getTopicContext().getRegion()));
+        tags.add(Tag.of(TAG_NAME_PROJECT, context.getTopicContext().getProjectName()));
+        tags.add(Tag.of(TAG_NAME_TOPIC, context.getTopicContext().getTopicName()));
+        tags.add(Tag.of(TAG_NAME_IDENTITY, context.getRequestContext().getProduceIdentity()));
+        tags.add(Tag.of(TAG_NAME_HOST, context.getRequestContext().getServiceHost()));
+        return tags;
+    }
+
+    private Counter getProducedBytesCounter(List<Tag> tags) {
+        // meter name - "produce.bytes", tags- region, project, topic, host
+        return Counter.builder(BYTES_METER).tags(tags).register(meterRegistry);
```

Comment:
do we have to register this counter every time whenever this method is called?
Having host in the metric tag/name has caused explosion in the total metrics in the past. We actually removed hosts from most places.

Reply:
https://micrometer.io/docs/concepts#_global_registry
This will return an existing meter if it was already registered. For now avoiding to cache these as well. Other option is to first find and then create if not found. That seems more expensive.
There is alternative method meterregistry.Counter(name, tags), it also does the same as above, might be more readable ?

Reply:
agree on host part might cause issues w.r.to cardinality, but it seems to be have value add for analysis purpose. With K8 it might become even more verbose. Not sure about it, currently it is set though.

Reply:
double check if opentelemetry is adding the host tag by default.